### PR TITLE
Gadget spawners

### PR DESCRIPTION
**Pull request by JanSchollaert**
# Description

Implemented gadget (weapon) spawners

I have a hard time changed the levels so I didn't change where the spawners are placed at. This will be done when refactoring the level anyway. (or when level editor is out and I can do it myself)

## Screenshot(s)
image

# How can this be tested?
Test with [59](https://github.com/Brick-Studios/BrickEngine/pull/59) (or master if that is merged)

- Can you pick up a weapon?
- If you empty the magazine or drop it off the map does it respawn?
- Is the weapon spawned random? (with the available_spawns from the level)
- When you pick up the weapon does it get placed at their hand / trunk (this is not perfect, pistol might be a bit off) <-- test this with bunny as well

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Included 1 or more screenshots
- [x] Code is const correct

### CODE REVIEW

File: lib/systems/gadget_spawn_system.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#include "brickengine/input.hpp"
+
+#include "systems/gadget_spawn_system.hpp"
+#include "components/spawn_component.hpp"
+#include "components/weapon_component.hpp"
+#include "brickengine/std/random.hpp"
+
+GadgetSpawnSystem::GadgetSpawnSystem(std::shared_ptr<CollisionDetector> cd,
+    std::shared_ptr<EntityManager> em, std::shared_ptr<EntityFactory> ef)
+    : BeastSystem(ef, em), collision_detector(cd) {}
```
Comment (MarkMeer) [inline lib/systems/gadget_spawn_system.cpp]: The collision detector is not used here and can be removed when not necessary in the future. I am not sure whether you both had some slick things for the collision detector in the gadget spawn system so I'll let it up to you ;-).

**Review by MarkMeer — COMMENTED** [review]
One small comment, but nothing major. Great work guys 👍 !

**Review by MarkMeer — APPROVED** [review]
Thank you Jan 👍

File: include/components/hold_component.hpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
+#ifndef FILE_GADGET_HOLDER_COMPONENT_HPP
```
Comment (Bram-Boris) [inline include/components/hold_component.hpp]: No gadget?

Comment (Bram-Boris) [inline include/components/hold_component.hpp]: No gadget?

File: include/systems/gadget_spawn_system.hpp [hunk 3]
```diff
@@ -0,0 +1,13 @@
+#ifndef FILE_GADGET_RESPAWN_SYSTEM_HPP
```
Comment (Bram-Boris) [inline include/systems/gadget_spawn_system.hpp]: reeeee

File: lib/components/spawn_component.cpp [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#include "components/spawn_component.hpp"
+
+SpawnComponent::SpawnComponent(int respawn_timer, std::vector<CreateGadgetCompsFn> gadget_fns) 
+                : respawn_timer(respawn_timer), time_elapsed(respawn_timer), gadget_fns(gadget_fns)  {}
+
+std::string SpawnComponent::SpawnComponent::getNameStatic() {
```
Comment (Bram-Boris) [inline lib/components/spawn_component.cpp]: Double class name?

File: lib/entities/entity_factory.cpp [hunk 5]
```diff
@@ -106,99 +182,49 @@ int EntityFactory::createElephant(int player_id) const {
     auto dst = std::unique_ptr<Rect>(new Rect{ 0, 0, 0, 0 });
     auto r = renderableFactory.createImage(GRAPHICS_PATH + "beasts/elephant/elephant-1.png", (int)Layers::Foreground, std::move(dst), 255);
     auto comps = std::make_unique<std::vector<std::unique_ptr<Component>>>();
-
+
```
Comment (Bram-Boris) [inline lib/entities/entity_factory.cpp]: ?

**Review by Bram-Boris — CHANGES_REQUESTED** [review]

**Review by MarkMeer — APPROVED** [review]
Agreed with the changes requested by Bram, not sure how I've missed those, but that's why we have two reviewers 😌.

**Review by Bram-Boris — APPROVED** [review]

**Pull request merged**